crates/red_knot_python_semantic/resources/mdtest/attributes.md

```diff
@@ -2,6 +2,212 @@
 
 Tests for attribute access on various kinds of types.
 
+## Class and instance variables
+
+### Pure instance variables
+
+#### Variable only declared/defined in `__init__`
+
+Variables only defined in `__init__` are pure instance variables. They can not be accessed on the
+class itself.
+
+```py
+class C:
+    def __init__(self, value2: int) -> None:
+        self.pure_instance_variable1 = "value set in __init__"
+        self.pure_instance_variable2 = value2
+        self.pure_instance_variable3: bytes
+
+c_instance = C(1)
+
+# TODO: should be `Literal["value set in __init__"]` (or `str` which would probably be more generally useful)
+reveal_type(c_instance.pure_instance_variable1)  # revealed: @Todo(instance attributes)
+
+# TODO: should be `int`
+reveal_type(c_instance.pure_instance_variable2)  # revealed: @Todo(instance attributes)
+
+# TODO: should be `bytes`
+reveal_type(c_instance.pure_instance_variable3)  # revealed: @Todo(instance attributes)
+
+c_instance.pure_instance_variable1 = "value set on instance"
+
+# TODO: this should be an error (incompatible types in assignment)
+c_instance.pure_instance_variable2 = "incompatible"
+
+# TODO: we already show an error here but the message might be improved?
+# mypy shows no error here, but pyright raises "reportAttributeAccessIssue"
+# error: [unresolved-attribute] "Type `Literal[C]` has no attribute `pure_instance_variable1`"
+reveal_type(C.pure_instance_variable1)  # revealed: Unknown
+
+# TODO: this should be an error (pure instance variables can not be accessed on the class)
+# mypy shows no error here, but pyright raises "reportAttributeAccessIssue"
+C.pure_instance_variable1 = "overwritten on class"
+
+# TODO: should ideally be `Literal["value set on instance"]`
+reveal_type(c_instance.pure_instance_variable1)  # revealed: @Todo(instance attributes)
+```
+
+#### Variable declared in class body and defined in `__init__`
+
+The same rule applies even if the variable is *declared* (not defined!) in the class body: it is
+still a pure instance variable.
+
+```py
+class C:
+    pure_instance_variable: str
+
+    def __init__(self) -> None:
+        self.pure_instance_variable = "value set in __init__"
+
+c_instance = C()
+
+# TODO: should be `Literal["value set in __init__"]` (or `str` which would probably be more generally useful)
```

Comment:
I think the explicit declaration in the class body here makes it unambiguous that the public type (the type seen from other scopes) should be `str` rather than a `Literal` type